akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala

```diff
@@ -0,0 +1,128 @@
+ * It offers additional hooks to modify the default behavior
+ */
+// @akka.annotation.ApiMayChange // FIXME replace with real annotation once Akka dependency bumped
+trait HttpApp {
+
+  private val serverBinding = new AtomicReference[ServerBinding]()
+
+  /**
+   * Start a server on the specified host and port, using provided settings
+   * Note that this method is blocking
+   */
+  def startServer(host: String, port: Int, settings: ServerSettings): Unit = {
+    startServer(host, port, settings, None)
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]]
+   * Note that this method is blocking
+   */
+  def startServer(host: String, port: Int, settings: ServerSettings, system: ActorSystem): Unit = {
+    startServer(host, port, settings, Some(system))
+  }
+
+  /**
+   * Start a server on the specified host and port, using provided settings and [[ActorSystem]] if present
+   * Note that this method is blocking
+   */
+  def startServer(host: String, port: Int, settings: ServerSettings, system: Option[ActorSystem]): Unit = {
+    implicit val theSystem = system.getOrElse(ActorSystem(Logging.simpleName(this).replaceAll("\\$", "")))
+    implicit val materializer = ActorMaterializer()
+    implicit val executionContext = theSystem.dispatcher
+
+    val bindingFuture = Http().bindAndHandle(
+      handler = route,
+      interface = host,
+      port = port,
+      settings = settings)
+
+    bindingFuture.onComplete {
+      case Success(binding) ⇒
+        //setting the server binding for possible future uses in the client
+        serverBinding.set(binding)
+        postHttpBinding(binding)
+      case Failure(cause) ⇒
+        postHttpBindingFailure(cause)
+    }
+
+    Await.ready(
+      bindingFuture.flatMap(_ ⇒ waitForShutdownSignal(theSystem)), // chaining both futures to fail fast
+      Duration.Inf) // It's waiting forever because maybe there is never a shutdown signal
+
+    bindingFuture
+      .flatMap(_.unbind())
+      .onComplete(attempt ⇒ {
+        postServerShutdown(attempt, theSystem)
+        // we created the system. we should cleanup!
+        if (system.isEmpty) theSystem.terminate()
+      })
+  }
+
+  /**
+   * It tries to retrieve the [[ServerBinding]] if the server has been successfully started. It fails otherwise.
+   * You can use this method to attemp to retrieve the [[ServerBinding]] at any point in time to, for example, stop the server due to unexpected circumstances.
+   */
+  def binding(): Try[ServerBinding] = {
+    if (serverBinding.get() == null) Failure(new IllegalStateException("Binding not yet stored. Have you called startServer?"))
+    else Success(serverBinding.get())
+  }
+
+  /**
+   * Hook that will be called just after the server termination. Override this method if you want to perform some cleanup actions after the server is stopped
+   */
+  protected def postServerShutdown(attempt: Try[Unit], system: ActorSystem): Unit = {
```

Comment:
Yeah and that's a mistake that we slipped through when making it stable sadly.
Let's use the proper types whenever we can, so Done here